In [35]:
import pip

In [36]:
!pip install dask[complete]

In [37]:
import dask.array as da
x = da.random.random((10000, 10000), chunks=(1000, 1000))
x

dask.array<random_sample, shape=(10000, 10000), dtype=float64, chunksize=(1000, 1000), chunktype=numpy.ndarray>

Cria dask array de dimensoes de 10000 por 10000

In [38]:
import dask.dataframe as dd

In [39]:
df = dd.read_csv('df2_completo.csv')

In [40]:
df.info()

<class 'dask.dataframe.core.DataFrame'>
Columns: 39 entries, order_id to product_width_cm
dtypes: object(23), float64(15), int64(1)

In [41]:
df_geolocation = dd.read_csv('olist_geolocation_dataset.csv')

In [42]:
df_geolocation.head(10)

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   
5                         1012       -23.547762       -46.635361   
6                         1047       -23.546273       -46.641225   
7                         1013       -23.546923       -46.634264   
8                         1029       -23.543769       -46.634278   
9                         1011       -23.547640       -46.636032   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP  
5        são paulo                SP  
6        sao paulo                SP  
7        sao paulo                SP  
8        sao paulo                SP  
9        sao paulo                SP

para um mesmo codigo postal tem-se lat e long diferentes, como por exemplo ali no 1046
vamos fazer entao a media para um mesmo codigo postal

In [43]:
#verificando dados faltantes
df_geolocation.isnull().compute()



geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                              False            False            False   
1                              False            False            False   
2                              False            False            False   
3                              False            False            False   
4                              False            False            False   
...                              ...              ...              ...   
1000158                        False            False            False   
1000159                        False            False            False   
1000160                        False            False            False   
1000161                        False            False            False   
1000162                        False            False            False   

         geolocation_city  geolocation_state  
0                   False              False  
1                   False              False  
2                   False              False  
3                   False              False  
4                   False              False  
...                   ...                ...  
1000158             False              False  
1000159             False              False  
1000160             False              False  
1000161             False              False  
1000162             False              False  

[1000163 rows x 5 columns]

In [44]:
df_geolocation.isnull().sum().compute()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [45]:
#teste
import pandas as pd
teste_geo = pd.read_csv("olist_geolocation_dataset.csv")

In [46]:
teste_geo.value_counts()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  geolocation_city  geolocation_state
88220                        -27.102099       -48.629613       itapema           SC                   314
6414                         -23.495901       -46.874687       barueri           SP                   189
                             -23.490618       -46.869004       barueri           SP                   127
5145                         -23.506049       -46.717377       sao paulo         SP                   126
22620                        -23.005514       -43.375964       rio de janeiro    RJ                   102
                                                                                                     ... 
18150                        -23.660997       -47.225822       ibiuna            SP                     1
                             -23.660702       -47.219962       ibiuna            SP                     1
                                                     

In [47]:
#quantidade de valores duplicados
#df_geolocation.value_counts().compute()

In [48]:
#utilizando o drop_duplicates, retorna um novo dataframe sem as duplicações
new_df_geo = df_geolocation.drop_duplicates()

In [49]:
new_df_geo.head(15)

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                          1037       -23.545621       -46.639292   
1                          1046       -23.546081       -46.644820   
2                          1046       -23.546129       -46.642951   
3                          1041       -23.544392       -46.639499   
4                          1035       -23.541578       -46.641607   
5                          1012       -23.547762       -46.635361   
6                          1047       -23.546273       -46.641225   
7                          1013       -23.546923       -46.634264   
8                          1029       -23.543769       -46.634278   
9                          1011       -23.547640       -46.636032   
10                         1013       -23.547325       -46.634184   
11                         1032       -23.538418       -46.634778   
12                         1014       -23.546435       -46.633830   
13                         1012       -23.548946       -46.634671   
14                         1037       -23.545187       -46.637855   

   geolocation_city geolocation_state  
0         sao paulo                SP  
1         sao paulo                SP  
2         sao paulo                SP  
3         sao paulo                SP  
4         sao paulo                SP  
5         são paulo                SP  
6         sao paulo                SP  
7         sao paulo                SP  
8         sao paulo                SP  
9         sao paulo                SP  
10        sao paulo                SP  
11        sao paulo                SP  
12        sao paulo                SP  
13        sao paulo                SP  
14        são paulo                SP

Agora, faz-se a media da latitude e longitude para o mesmo cep

In [50]:
#utilizando groupby

new_df_geo = new_df_geo.groupby(by="geolocation_zip_code_prefix").mean()


In [51]:
new_df_geo.head(15)

geolocation_lat  geolocation_lng
geolocation_zip_code_prefix                                  
1001                              -23.550227       -46.634039
1002                              -23.547657       -46.634991
1003                              -23.549000       -46.635582
1004                              -23.549829       -46.634792
1005                              -23.549547       -46.636406
1006                              -23.550127       -46.636044
1007                              -23.549962       -46.637204
1008                              -23.546000       -46.635877
1009                              -23.546891       -46.636498
1010                              -23.546667       -46.635404
1011                              -23.546649       -46.635425
1012                              -23.547808       -46.634844
1013                              -23.547240       -46.634221
1014                              -23.546040       -46.633600
1015                              -23.547756       -46.631738

Para cada vendedor e para cada cliente ter um cep associado

DataFrame.merge(right, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, suffixes=('_x', '_y'), indicator=False, npartitions=None, shuffle=None, broadcast=None)[source]

Merge df1 and df2 on the lkey and rkey columns. The value columns have the default suffixes, _x and _y, appended
-> df1.merge(df2, left_on='lkey', right_on='rkey')


In [52]:
#DataFrame.merge(right, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, suffixes=('_x', '_y'), indicator=False, npartitions=None, shuffle=None, broadcast=None)[source]

#Merge df1 and df2 on the lkey and rkey columns. The value columns have the default suffixes, _x and _y, appended
#df1.merge(df2, left_on='lkey', right_on='rkey')

In [53]:
#teste = df.merge(df,how='inner',on='costumer_zip_code_prefix')
#como nao possuem colunas iguais, e preciso identificar qual a coluna de cada df deve ser utilizada como referencia, por isso aqui o on nao funciona

In [54]:
#Identificando o cliente
#left para o df completo e rigth para o com as localizações
costumer = df.merge(new_df_geo,left_on='customer_zip_code_prefix', right_on='geolocation_zip_code_prefix')

In [55]:
#certo, cada cliente tem um lat e um long
costumer.head(5)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214            1.0   
1  bd31b009e1dbc47fc7c250b1e2cf5440            1.0   
2  bb4eb0196897c20281a61f75ce23211c            1.0   
3  f01cc1e4fe1833e42f696b0dcb0d77ef            1.0   
4  f4776a61a23801c349ab69bd0fd774e2            1.0   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  92e2d2146e433cd4d1f09a3f8633ead0  4a3ca9315b744ce9f8e9374361493884   
2  8cee1d824765335c48ccc515c2ecf4b8  4a3ca9315b744ce9f8e9374361493884   
3  281886d5650643134700ec611ac33532  218d46b86c1881d022bce9c68a7d4b15   
4  5a6e53c3b4e8684b13388d6aa4afdf12  7299e27ed73d2ad986de7f7c77d919fa   

   shipping_limit_date   price  freight_value  payment_sequential  \
0  2017-09-19 09:45:35   58.90          13.29                 1.0   
1  2018-05-02 03:51:18   52.90          22.95                 1.0   
2  2017-07-28 14:50:11   96.00          15.42                 1.0   
3  2017-11-20 19:55:23  220.00          18.79                 1.0   
4  2017-05-23 04:02:17   16.99          15.10                 1.0   

  payment_type  payment_installments  ...  product_category_name  \
0  credit_card                   2.0  ...             cool_stuff   
1       boleto                   1.0  ...        cama_mesa_banho   
2  credit_card                   9.0  ...        cama_mesa_banho   
3  credit_card                   3.0  ...          esporte_lazer   
4       boleto                   1.0  ...  utilidades_domesticas   

  product_name_lenght  product_description_lenght product_photos_qty  \
0                58.0                       598.0                4.0   
1                53.0                       714.0                1.0   
2                54.0                       385.0                2.0   
3                31.0                      1327.0                1.0   
4                45.0                       176.0                2.0   

  product_weight_g product_length_cm product_height_cm product_width_cm  \
0            650.0              28.0               9.0             14.0   
1           1200.0              33.0              18.0             26.0   
2            900.0              30.0               9.0             20.0   
3           1100.0              16.0              16.0             16.0   
4            200.0              19.0               9.0             12.0   

  geolocation_lat geolocation_lng  
0      -21.763186      -41.310265  
1      -21.763186      -41.310265  
2      -21.763186      -41.310265  
3      -21.763186      -41.310265  
4      -21.763186      -41.310265  

[5 rows x 41 columns]

Problema: agora tem-se duas colunas com as informações de geolocalizacao
retira-se as erradas: costumer_lat e costumer_long

In [56]:
#a coluna para latitude do cliente recebe o valor correto
costumer['costumer_lat'] = costumer['geolocation_lat']

In [57]:
#apaga a variavel geolocation que nao precisa mais
costumer = costumer.drop(['geolocation_lat'],axis=1)

In [58]:
#a coluna para longitude do cliente recebe o valor correto
costumer['costumer_lng'] = costumer['geolocation_lng']

In [59]:
#apaga a variavel geolocation que nao precisa mais
costumer = costumer.drop(['geolocation_lng'],axis=1)

In [60]:
costumer.head(5)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214            1.0   
1  bd31b009e1dbc47fc7c250b1e2cf5440            1.0   
2  bb4eb0196897c20281a61f75ce23211c            1.0   
3  f01cc1e4fe1833e42f696b0dcb0d77ef            1.0   
4  f4776a61a23801c349ab69bd0fd774e2            1.0   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  92e2d2146e433cd4d1f09a3f8633ead0  4a3ca9315b744ce9f8e9374361493884   
2  8cee1d824765335c48ccc515c2ecf4b8  4a3ca9315b744ce9f8e9374361493884   
3  281886d5650643134700ec611ac33532  218d46b86c1881d022bce9c68a7d4b15   
4  5a6e53c3b4e8684b13388d6aa4afdf12  7299e27ed73d2ad986de7f7c77d919fa   

   shipping_limit_date   price  freight_value  payment_sequential  \
0  2017-09-19 09:45:35   58.90          13.29                 1.0   
1  2018-05-02 03:51:18   52.90          22.95                 1.0   
2  2017-07-28 14:50:11   96.00          15.42                 1.0   
3  2017-11-20 19:55:23  220.00          18.79                 1.0   
4  2017-05-23 04:02:17   16.99          15.10                 1.0   

  payment_type  payment_installments  ...  product_category_name  \
0  credit_card                   2.0  ...             cool_stuff   
1       boleto                   1.0  ...        cama_mesa_banho   
2  credit_card                   9.0  ...        cama_mesa_banho   
3  credit_card                   3.0  ...          esporte_lazer   
4       boleto                   1.0  ...  utilidades_domesticas   

  product_name_lenght  product_description_lenght product_photos_qty  \
0                58.0                       598.0                4.0   
1                53.0                       714.0                1.0   
2                54.0                       385.0                2.0   
3                31.0                      1327.0                1.0   
4                45.0                       176.0                2.0   

  product_weight_g product_length_cm product_height_cm product_width_cm  \
0            650.0              28.0               9.0             14.0   
1           1200.0              33.0              18.0             26.0   
2            900.0              30.0               9.0             20.0   
3           1100.0              16.0              16.0             16.0   
4            200.0              19.0               9.0             12.0   

  costumer_lat costumer_lng  
0   -21.763186   -41.310265  
1   -21.763186   -41.310265  
2   -21.763186   -41.310265  
3   -21.763186   -41.310265  
4   -21.763186   -41.310265  

[5 rows x 41 columns]

Para o vendedor

In [61]:
seller = costumer.merge(new_df_geo,left_on='seller_zip_code_prefix', right_on='geolocation_zip_code_prefix')

In [62]:
#apaga a variavel geolocation que nao precisa mais
#a coluna para latitude do vendedor recebe o valor correto
seller['seller_lat'] = seller['geolocation_lat']
seller = seller.drop(['geolocation_lat'],axis=1)

In [63]:
#apaga a variavel geolocation que nao precisa mais
#a coluna para long do vendedor recebe o valor correto
seller['seller_lng'] = seller['geolocation_lng']
seller = seller.drop(['geolocation_lng'],axis=1)

In [64]:
seller.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value', 'payment_sequential',
       'payment_type', 'payment_installments', 'payment_value', 'review_id',
       'review_score', 'review_comment_title', 'review_comment_message',
       'review_creation_date', 'review_answer_timestamp', 'customer_id',
       'order_status', 'order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'seller_zip_code_prefix', 'seller_city', 'seller_state',
       'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'costumer_lat', 'costumer_lng', 'seller_lat', 'seller_lng'],
      dtype='object')

In [65]:
costumer.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value', 'payment_sequential',
       'payment_type', 'payment_installments', 'payment_value', 'review_id',
       'review_score', 'review_comment_title', 'review_comment_message',
       'review_creation_date', 'review_answer_timestamp', 'customer_id',
       'order_status', 'order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'seller_zip_code_prefix', 'seller_city', 'seller_state',
       'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'costumer_lat', 'costumer_lng'],
      dtype='object')

In [67]:
seller = seller.compute()

In [68]:
seller.to_csv('olist_geolocated_transactions.csv')